### DISCUSSION TITLE
really phalcon 4 is stable?

### DISCUSSION THREAD

**Discussion by garag-lib** [phalcon/cphalcon#15926] [General]
just creating the model with phalcon-devtools findfirst generates exception

must implement interface Phalcon\Mvc\ModelInterface or be null, instance of Phalcon\Mvc\Model\Row returned

that the basic works does not work

wasting time migrating is stupid

**Comment by BeMySlaveDarlin** [comment]
Phalcon 4 requires devtools 4 version as well. 
Framework itself is production ready.
Helper libraries - partly outdated. Feel free to make pull requests to make libraries more relevant to framework.

**Comment by garag-lib** [comment]
simply creating a model with devtool 4 generates an exception

example:

    /**
     * Allows to query the first record that match the specified conditions
     *
     * @param mixed $parameters
     * @return Logins|\Phalcon\Mvc\Model\ResultInterface|\Phalcon\Mvc\ModelInterface|null
     */
    public static function findFirst($parameters = null): ?\Phalcon\Mvc\ModelInterface {
        return parent::findFirst($parameters);
    }

this generates exception

must implement interface Phalcon\Mvc\ModelInterface or be null, instance of Phalcon\Mvc\Model\Row returned

This way we cant work, 4 is not stable

if you don't overload findfirst work, ok. But this is horrible, becasuse your ide cant work with annotations.

this is crappy

phalcon 4 is not for production

**Reply by BeMySlaveDarlin** [reply]
Create issue to devtools repository, ill create fix asap.
With some examples of your generated code, pls.

**Reply by garag-lib** [reply]
it's not a devtools bug, it is a framework error. 

I understand beta versions. Phalcon 4 not is production version.

**Reply by BeMySlaveDarlin** [reply]
Can you provide real example of code, producing that exception?

**Comment by garag-lib** [comment]
model:

`
namespace minamespace;

/**
 * Logins
 * 
 * @package example
 * @autogenerated by Phalcon Developer Tools
 * @date 2022-03-20, 18:59:33
 */
class Logins extends \Phalcon\Mvc\Model {

    /**
     *
     * @var string
     */
    public $user;

    /**
     *
     * @var string
     */
    public $token;

    /**
     *
     * @var string
     */
    public $fecha;

    /**
     *
     * @var string
     */
    public $info;

    /**
     * Initialize method for model.
     */
    public function initialize() {
        $this->setSource("logins");
    }

    /**
     * Allows to query a set of records that match the specified conditions
     *
     * @param mixed $parameters
     * @return Logins[]|Logins|\Phalcon\Mvc\Model\ResultSetInterface
     */
    public static function find($parameters = null): \Phalcon\Mvc\Model\ResultsetInterface {
        return parent::find($parameters);
    }

    /**
     * Allows to query the first record that match the specified conditions
     *
     * @param mixed $parameters
     * @return Logins|\Phalcon\Mvc\Model\ResultInterface|\Phalcon\Mvc\ModelInterface|null
     */
    public static function findFirst($parameters = null): ?\Phalcon\Mvc\ModelInterface {
        return parent::findFirst($parameters);
    }

    /**
     * Independent Column Mapping.
     * Keys are the real names in the table and the values their names in the application
     *
     * @return array
     */
    public function columnMap() {
        return static::$_columnMap;
    }

    public static $_columnMap = [
        'user' => 'user',
        'token' => 'token',
        'fecha' => 'fecha',
        'info' => 'info'
    ];

}
`

one model

`        $example = \minamespace\Logins::findFirst([
                    'columns' => 'user'
        ]);`

I must put columns, becasuse if you put  $example = \minamespace\Logins::findFirst(); the sql generate it's this: SELECT FROM "Logins" LIMIT :APL0 APL0 => 1 and generates error

ok, exception:

ERROR: Return value of minamespace\Logins::findFirst() must implement interface Phalcon\Mvc\ModelInterface or be null, instance of Phalcon\Mvc\Model\Row returned
# function: findFirst
file: /var/www/example/app/controllers/IndexController.php
line: 23
class: crononet\Usuarios
# function: indexAction
class: IndexController
# function: callActionMethod
class: Phalcon\Dispatcher\AbstractDispatcher
# function: dispatch
file: /var/www/manuelpeliz.es/public/kernel.php
line: 110
class: Phalcon\Dispatcher\AbstractDispatcher


a horror

**Comment by garag-lib** [comment]
the code generated by devtools is correct, but the framework has serious errors

**Comment by garag-lib** [comment]
https://github.com/phalcon/cphalcon/issues/15212

If someone wants to migrate, they should be told not to do so. 

with these problems can not work

**Comment by BeMySlaveDarlin** [comment]
Tested with fresh install, I approve this is bug, but related directly to `columns` part of conditions.
When you use `column` select, `Query` cannot create model itself (model requires full fields fetched) and returns scalar or row object.
As i see, `columns` condition cannot be and should not be used via Models paradigm, `findFirst` expects `where` conditions.

To get filtered columns or scalar values from database, use query builder (phql) instead.

**Reply by garag-lib** [reply]
that would be great if "$example = \minamespace\Logins::findFirst()" it works. but no

pd: sqlite at least
